libobs, win-wasapi: Enable channel routing for Audio Input Capture

pkviet opened the pull request:
### Description
On libobs side, this adds the capability for the audio resampler (relying on swresampler lib) to
set a matrix for routing of channels.
On win-wasapi side, this applies the new capability to enable channel selection.
This brings parity to wasapi input capture with coreaudio input capture on macOS.

#### Screenshots

<img width="840" height="619" alt="obs64_2025-12-17_17-35-57" src="https://github.com/user-attachments/assets/f80697fa-0c53-456a-bf59-da884a61e93e" />

<img width="840" height="619" alt="obs64_2025-12-17_17-36-14" src="https://github.com/user-attachments/assets/8f127ba9-6dec-46a6-899e-955f840873d8" />

### Motivation and Context
One motivation for that feature are devices which separate stereo inputs.
So if a mike is connected to , say, Input 1 while nothing is connected to Input 2, wasapi usually treats them as
stereo pairs and when obs audio is set to Stereo, the captured Right channel will be silent.
With channel routing capability, it is now possible to duplicate the left channel to the right and have a consistent level.
The issue was brought to my attention by @Fenrirthviti 


### How Has This Been Tested?
Tested on windows 11 pro 25H2 w/ a logitech webcam + USB sound board.
Disabling the downmix cehckbox enables channel routing.
Swapping between devices resets the channel selections.
Changing the channel selections updates the captured channel.

### Types of changes
- New feature (non-breaking change which adds functionality) 

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

pkviet committed:
libobs: Enable audio resampler to set a custom channel matrix

This adds a function allowing to set a custom channel matrix, allowing
channel rerouting. This is useful when sources provide layouts not
supported by obs or to correct incorrect channel assignments.

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
win-wasapi: Enable channel selections

This adds the capability to select input channels exposed to obs.
(The same is already possible for coreaudio input sources.)

Signed-off-by: pkv <pkv@obsproject.com>

pkviet committed:
mac-capture: Add hint for downmix checkbox of coreaudio input

This adds a hint for the 'downmix' checkbox.
It is likely unclear to users what this checkbox does or does not.
When disabled, it allows channel selection; when enabled it lets obs
deal with speaker layout mismatches between source and obs.

Signed-off-by: pkv <pkv@obsproject.com>

PatTheMav commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -481,6 +512,29 @@ WASAPISource::UpdateParams WASAPISource::BuildUpdateParams(obs_data_t *settings)
+				obs_data_set_int(settings, channel_str.c_str(), -1);
+				params.mixChannels[i] = -1;
+			}
+			resamplerMatrixNeedsUpdate.store(true, std::memory_order_relaxed);
```
The relaxed memory order here means that the CPU (and potentially compiler) can move setting `resamplerMatrixNeedsUpdate` to `true` before the loop above has set the `mixChannels`. If that's ok (because they are unrelated), no changes are needed.

PatTheMav reviewed (commented):
Looks fine to me overall, also tested this with an virtual 6-ch audio device and allowed me to map them as expected.

Haven't done in-depth checks whether the audio is correctly routed though.

pkviet commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -304,6 +319,14 @@ WASAPISource::WASAPISource(obs_data_t *settings, obs_source_t *source_, SourceTy
+	resampler = nullptr;
+	deviceUpdated.store(false, std::memory_order_relaxed);
+	resamplerData[0] = nullptr;
+	resamplerMatrixNeedsUpdate.store(true, std::memory_order_release);
```
resamplerMatrixNeedsUpdate is set to true when:
- the wasapi source is created (line 325),
- when the device is swapped to another or when the downmix checkbox is changed to false (line 534) (in UI thread)
- when the user changes the mixChannels in Properties window in UI thread,
It is swapped to false when the resampler is updated in capture thread (line 1149)

pkviet replied:
mmh you're right in principle but there is no risk.
So the consumer is the capture thread. It is not started before a first round in the UI thread of params updates. (UpdateSettings is called in the CTOR). So the capture thread won't process incorrect or unset values of mixChannels.
For full correctness since the bool tracks mixChannels, it should be committed after the block changing it. I have updated the code to reflect that.

RytoEX commented on plugins/mac-capture/data/locale/en-US.ini:
```diff
@@ -8,6 +8,7 @@ CoreAudio.Channel.Unnamed="Unnamed"
 CoreAudio.Channel.Device="Device Channel"
 CoreAudio.None="None"
 CoreAudio.Downmix="Enable Downmixing"
+CoreAudio.Downmix.Hint="Unchecked: this allows to control channel routing to obs.\nChecked (default): obs will automatically upmix or downmix to obs speaker layout."
```
Always capitalize OBS in user-facing text.
```suggestion
CoreAudio.Downmix.Hint="Unchecked: this allows to control channel routing to OBS.\nChecked (default): obs will automatically upmix or downmix to OBS speaker layout."
```

Not sure about the wording here or the order of Unchecked vs. Checked descriptions. cc @Warchamp7

pkviet replied:
capitalized. Resolving until there is further guidance on the wording

RytoEX commented on plugins/win-wasapi/data/locale/en-US.ini:
```diff
@@ -9,3 +9,14 @@ Priority="Window Match Priority"
 Priority.Class="Match title, otherwise find window of same type"
 Priority.Exe="Match title, otherwise find window of same executable"
+Downmix="Enable Downmixing"
+Downmix.Hint="Unchecked: this allows to control channel routing to obs.\nChecked (default): obs will automatically upmix or downmix to obs speaker layout."
```
```suggestion
Downmix.Hint="Unchecked: this allows to control channel routing to OBS.\nChecked (default): obs will automatically upmix or downmix to OBS speaker layout."
```

See also comment in `plugins/mac-capture/data/locale/en-US.ini`

pkviet replied:
fixed

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -13,6 +13,7 @@
 #include <util/windows/window-helpers.h>
 #include <util/threading.h>
 #include <util/util_uint64.h>
+#include <media-io/audio-resampler.h>
```
I don't recall if we give `obs-module.h` and/or `obs.h` special treatment when ordering includes, but this new include should at least be sorted with the rest of the `util/*` includes. cc @PatTheMav for clarification.

PatTheMav replied:
Applying the rules as intended would generate this include block:

```C++
#include <obs-module.h>
#include <obs.h>
#include <util/dstr.h>
#include <util/platform.h>
#include <util/threading.h>
#include <util/util_uint64.h>
#include <util/windows/CoTaskMemPtr.hpp>
#include <util/windows/ComPtr.hpp>
#include <util/windows/HRError.hpp>
#include <util/windows/WinHandle.hpp>
#include <util/windows/win-version.h>
#include <util/windows/window-helpers.h>
```

I usually select the lot of them in Sublime Text and use "Sort Lines (Case Sensitive)" and that's it.

---

I'm less sure about the C++ STL and Win32 headers - mashing them together (as I suggested off-thread with regards to `windows.h`) makes a bit of a mess, and visually I find this variant to be more expressive:

```C++
#include <RTWorkQ.h>
#include <audioclientactivationparams.h>
#include <avrt.h>
#include <wrl/implements.h>

#include <atomic>
#include <cinttypes>
```
Win32 headers in in one block as "system" provided includes first, followed by cross-platform C++ STL headers (the same would IMO then also apply to Cocoa/CoreFoundation includes before cross-platform headers on macOS).

pkviet replied:
put #include <media-io/audio-resampler.h> in alphabetical order.
Sorted the other includes in the order suggested by Pat.

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -517,6 +571,18 @@ void WASAPISource::UpdateSettings(UpdateParams &&params)
+		bool changed = false;
+		for (int i = 0; i < MAX_AUDIO_CHANNELS; i++) {
+			if (mixChannels[i] != params.mixChannels[i])
+				changed = true;
```
```suggestion
			if (mixChannels[i] != params.mixChannels[i]) {
				changed = true;
			}
```

pkviet replied:
done

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -517,6 +571,18 @@ void WASAPISource::UpdateSettings(UpdateParams &&params)
+			mixChannels[i] = std::move(params.mixChannels[i]);
+		}
+		if (changed)
+			resamplerMatrixNeedsUpdate.store(true, std::memory_order_release);
```
```suggestion
		if (changed) {
			resamplerMatrixNeedsUpdate.store(true, std::memory_order_release);
		}
```

pkviet replied:
done

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -815,6 +893,34 @@ ComPtr<IAudioCaptureClient> WASAPISource::InitCapture(IAudioClient *client, HAND
+void WASAPISource::UpdateResamplerMatrix()
+{
+	if (!resampler)
+		return;
```
```suggestion
	if (!resampler) {
		return;
	}
```

pkviet replied:
fixed

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -815,6 +893,34 @@ ComPtr<IAudioCaptureClient> WASAPISource::InitCapture(IAudioClient *client, HAND
+	const int in_ch = inputChannels;
+
+	if (!in_ch)
+		return;
```
```suggestion
	if (!in_ch) {
		return;
	}
```

pkviet replied:
fixed

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -1012,9 +1144,33 @@ bool WASAPISource::ProcessCaptureData()
-		data.data[0] = buffer;
+		if (sourceType == SourceType::Input) {
+			if (resamplerMatrixNeedsUpdate.exchange(false, std::memory_order_acq_rel))
+				UpdateResamplerMatrix();
```
```suggestion
			if (resamplerMatrixNeedsUpdate.exchange(false, std::memory_order_acq_rel)) {
				UpdateResamplerMatrix();
			}
```

pkviet replied:
fixed

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -1461,10 +1623,85 @@ static bool UpdateWASAPIMethod(obs_properties_t *props, obs_property_t *, obs_da
+	UNUSED_PARAMETER(props);
+	WASAPISource *data = (WASAPISource *)vptr;
+	if (data == NULL)
+		return false;
```
```suggestion
	if (data == NULL) {
		return false;
	}
```

pkviet replied:
fixed

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -1461,10 +1623,85 @@ static bool UpdateWASAPIMethod(obs_properties_t *props, obs_property_t *, obs_da
+
+	const char *device_id = obs_data_get_string(settings, OPT_DEVICE_ID);
+	const char *device_name = "";
+	vector<AudioDeviceInfo> devices;
```
```suggestion
	std::vector<AudioDeviceInfo> devices;
```

pkviet replied:
fixed

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -1461,10 +1623,85 @@ static bool UpdateWASAPIMethod(obs_properties_t *props, obs_property_t *, obs_da
 {
+	WASAPISource *data = (WASAPISource *)vptr;
+	if (data == NULL)
+		return false;
```
```suggestion
	if (data == NULL) {
		return false;
	}
```

pkviet replied:
fixed

RytoEX replied:
This one seems unfixed.

pkviet replied:
ah sorry, fixed now

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -1461,10 +1623,85 @@ static bool UpdateWASAPIMethod(obs_properties_t *props, obs_property_t *, obs_da
+		if (enable)
+			obs_data_set_int(settings, channel_str.c_str(), -1);
+		obs_property_set_modified_callback2(mix_channels[i], channel_update_cb, data);
+		channel_update_cb(data, props, mix_channels[i], settings);
```
```suggestion
		if (enable) {
			obs_data_set_int(settings, channel_str.c_str(), -1);
		}
		obs_property_set_modified_callback2(mix_channels[i], channel_update_cb, data);
		channel_update_cb(data, props, mix_channels[i], settings);
```

pkviet replied:
fixed

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -1461,10 +1623,85 @@ static bool UpdateWASAPIMethod(obs_properties_t *props, obs_property_t *, obs_da
+{
+	WASAPISource *data = (WASAPISource *)vptr;
+	if (data == NULL)
+		return false;
```
```suggestion
	if (data == NULL) {
		return false;
	}
```

pkviet replied:
fixed

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -1461,10 +1623,85 @@ static bool UpdateWASAPIMethod(obs_properties_t *props, obs_property_t *, obs_da
+{
+	WASAPISource *data = (WASAPISource *)vptr;
+	if (data == NULL)
+		return false;
```
```suggestion
	if (data == NULL) {
		return false;
	}
```

pkviet replied:
fixed

RytoEX commented on plugins/win-wasapi/win-wasapi.cpp:
```diff
@@ -1479,8 +1716,33 @@ static obs_properties_t *GetWASAPIPropertiesInput(void *)
 	}
 
+	if (data != nullptr)
+		obs_property_set_modified_callback2(device_prop, device_update_cb, data);
```
```suggestion
	if (data != nullptr) {
		obs_property_set_modified_callback2(device_prop, device_update_cb, data);
	}
```

pkviet replied:
fixed

RytoEX reviewed (changes requested):
cc @Warchamp7 on brace changes. If needed, we can just update all of these in the clang-format PR, but we will soon be requiring and enforcing braces for all control statements in C++.

Warchamp7 reviewed (changes requested):
A couple bits of feedback and some questions.

Can we invert the checkbox?
i.e. Unchecked by default, and the label is instead "Enable custom channel routing"

---

When `Device` is set to `Default` the channel selection dropdowns are missing a device name and simply say ` channel 1`, etc.

<img width="437" height="330" alt="Image" src="https://github.com/user-attachments/assets/6891c6d5-d0ff-41a9-b0a0-233e20b84c85" />

Ideally the actual device name should be used in the device channel dropdowns and not just `Default`.

---

At time of review, wasapi output devices don't appear to be working:
```
17:18:30.101: [win-wasapi: 'Audio Output Capture'] update settings:
17:18:30.101: 	device id: {0.0.0.00000000}.{825bd69f-078a-481c-831c-eda29a7002bb}
17:18:30.101: 	use device timing: 1
17:18:30.108: Device 'Speakers (Scarlett 2i2 4th Gen)' invalidated.  Retrying (source: Audio Output Capture)
17:18:30.118: warning: Input channel layout "" is invalid or unsupported.
17:18:30.118: avresample_open failed: error code -22
17:18:30.118: WASAPI: Device 'Speakers (Realtek USB2.0 Audio)' [48000 Hz] initialized (source: Audio Output Capture)```

pkviet commented:
> A couple bits of feedback and some questions.
> 
> Can we invert the checkbox? i.e. Unchecked by default, and the label is instead "Enable custom channel routing"
> 

This can be done, but this means wasapi would not be aligned with core audio where the feature was first implemented.

> When `Device` is set to `Default` the channel selection dropdowns are missing a device name and simply say ` channel 1`, etc.
> Ideally the actual device name should be used in the device channel dropdowns and not just `Default`.
> 

Ok I will add the logic for that.

> At time of review, wasapi output devices don't appear to be working:
> 

Indeed. My scope was to bring parity to wasapi with coreaudio.
I didn't consider Audio Output Capture.
I'll investigate the extent of changes that would be required for that and either add that or defer to another PR.

Warchamp7 commented:
> > A couple bits of feedback and some questions.
> > Can we invert the checkbox? i.e. Unchecked by default, and the label is instead "Enable custom channel routing"
> 
> This can be done, but this means wasapi would not be aligned with core audio where the feature was first implemented.

I'd like to update core audio to match that as well then. It can be a separate PR though.

pkviet commented:
**Update**
Addressed @Warchamp7 comments.

1. Changed UI so that the checkbox now enables channel routing ; checkbox unchecked:
<img width="882" height="236" alt="obs64_pZWnelHwtl" src="https://github.com/user-attachments/assets/8fd2602d-d01f-4d23-a4cc-1b396db0b98b" />    
     

checkbox checked:
<img width="833" height="278" alt="obs64_IiMGxFsxQn" src="https://github.com/user-attachments/assets/acc83cfb-d912-4a17-bb04-ec4640647cf2" />

2.  Default device now displays the device name in channels list:
<img width="840" height="583" alt="obs64_ZVGYtk3YO9" src="https://github.com/user-attachments/assets/7981e583-7f6e-49ab-b0e0-b3d55c0ccc9a" />

3. Fixed the issue where Audio Output Capture wouldn't capture anymore.

Warchamp7 commented on plugins/win-wasapi/data/locale/en-US.ini:
```diff
@@ -9,3 +9,14 @@ Priority="Window Match Priority"
 Priority.Class="Match title, otherwise find window of same type"
 Priority.Exe="Match title, otherwise find window of same executable"
+Routing="Enable Channel Routing"
+Routing.Hint="Checked: this allows to control channel routing to OBS.\nUnchecked (default): OBS will automatically upmix or downmix to OBS speaker layout."
```
```suggestion
Routing.Hint="By default OBS will automatically upmix or downmix to the number of channels configured in settings. Enabling this setting will allow you to manually select which channels of the device to map to the input in OBS"
```

pkviet replied:
Done.

<img width="1122" height="277" alt="obs64_Dr6WX2amTr" src="https://github.com/user-attachments/assets/d29349ac-ab8c-491f-b5cf-d1e42a4aba58" />
